apoc.export.cypher.all fails when a compound constraint exists

## Expected Behavior (Mandatory)
`CALL apoc.export.cypher.all('/data/seed.cypher',{format:'cypher-shell'});` works when compound constraint exist.

## Actual Behavior (Mandatory)
```
CALL apoc.export.cypher.all('/data/seed.cypher',{format:'cypher-shell'});
Failed to invoke procedure `apoc.export.cypher.all`: Caused by: java.util.NoSuchElementException: More than one element in java.util.AbstractList$Itr@693ec7a0. First element is 'adTagType' and the second element is 'name'
```

## How to Reproduce the Problem
Create a compound constraint on a node. I have the following:

```
create constraint on (t:AdTag) assert (t.adTagType, t.name) is node key;
```

### Simple Dataset (where it's possibile)

```
MERGE (a:AdTag:test {name: "E", adTagType: "ESRB-test"});
MERGE (a:AdTag:test {name: "Under 18", adTagType: "Age-test"});
MERGE (a:AdTag:test {name: "13-17",  adTagType: "Age-test"});
MERGE (a:AdTag:test {name: "18-24", adTagType: "Age-test"});
MERGE (a:AdTag:test {name: "25-34", adTagType: "Age-test"});
```


### Steps (Mandatory)

  1. Run `create constraint on (t:AdTag) assert (t.adTagType, t.name) is node key;`
  1. Add the sample data above.
  1. Run `CALL apoc.export.cypher.all('/data/seed.cypher',{format:'cypher-shell'});`

## Screenshots (where it's possibile)

## Specifications (Mandatory)

Currently used versions

### Versions

  - OS: OS X 10.13.6
  - Neo4j: Enterprise 3.5.0
  - Neo4j-Apoc: 3.5.0.1

--- a/src/main/java/apoc/export/cypher/MultiStatementCypherSubGraphExporter.java
+++ b/src/main/java/apoc/export/cypher/MultiStatementCypherSubGraphExporter.java
@@ -1,5 +1,6 @@
 package apoc.export.cypher;
 
+import apoc.export.cypher.formatter.CypherFormatterUtils;
 import apoc.export.util.*;
 import apoc.export.cypher.formatter.CypherFormatter;
 import apoc.util.Util;
@@ -13,6 +14,8 @@ import org.neo4j.helpers.collection.Iterables;
 import java.io.IOException;
 import java.io.PrintWriter;
 import java.util.*;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
 
 import static apoc.export.cypher.formatter.CypherFormatterUtils.*;
 
@@ -26,7 +29,7 @@ import static apoc.export.cypher.formatter.CypherFormatterUtils.*;
 public class MultiStatementCypherSubGraphExporter {
 
     private final SubGraph            graph;
-    private final Map<String, String> uniqueConstraints = new HashMap<>();
+    private final Map<String, Set<String>> uniqueConstraints = new HashMap<>();
     private Set<String> indexNames        = new LinkedHashSet<>();
     private Set<String> indexedProperties = new LinkedHashSet<>();
     private Long artificialUniques = 0L;
@@ -140,7 +143,7 @@ public class MultiStatementCypherSubGraphExporter {
             out.println(index);
         }
         if (artificialUniques > 0) {
-            String cypher = this.cypherFormat.statementForConstraint(UNIQUE_ID_LABEL, UNIQUE_ID_PROP);
+            String cypher = this.cypherFormat.statementForConstraint(UNIQUE_ID_LABEL, Collections.singleton(UNIQUE_ID_PROP));
             if (cypher != null && !"".equals(cypher)) {
                 out.println(cypher);
             }
@@ -160,7 +163,8 @@ public class MultiStatementCypherSubGraphExporter {
             String label = index.getLabel().name();
             Iterable<String> props = index.getPropertyKeys();
             if (index.isConstraintIndex()) {
-                String cypher = this.cypherFormat.statementForConstraint(label, Iterables.single(props));
+                String cypher;
+                cypher = this.cypherFormat.statementForConstraint(label, props);
                 if (cypher != null && !"".equals(cypher)) {
                     result.add(cypher);
                 }
@@ -199,7 +203,7 @@ public class MultiStatementCypherSubGraphExporter {
                 artificialUniques -= batchSize;
             }
             begin(out);
-            String cypher = this.cypherFormat.statementForConstraint(UNIQUE_ID_LABEL, UNIQUE_ID_PROP).replaceAll("^CREATE", "DROP");
+            String cypher = this.cypherFormat.statementForConstraint(UNIQUE_ID_LABEL, Collections.singleton(UNIQUE_ID_PROP)).replaceAll("^CREATE", "DROP");
             if (cypher != null && !"".equals(cypher)) {
                 out.println(cypher);
             }
@@ -230,11 +234,12 @@ public class MultiStatementCypherSubGraphExporter {
     private void gatherUniqueConstraints() {
         for (IndexDefinition indexDefinition : graph.getIndexes()) {
             String label = indexDefinition.getLabel().name();
-            String prop = Iterables.first(indexDefinition.getPropertyKeys());
+            //String prop = Iterables.first(indexDefinition.getPropertyKeys());
+            Set<String> props = StreamSupport.stream(indexDefinition.getPropertyKeys().spliterator(), false).collect(Collectors.toSet());
             indexNames.add(label);
-            indexedProperties.add(prop);
+            indexedProperties.addAll(props);
             if (indexDefinition.isConstraintIndex()) {
-                if (!uniqueConstraints.containsKey(label)) uniqueConstraints.put(label, prop);
+                if (!uniqueConstraints.containsKey(label)) uniqueConstraints.put(label, props);
             }
         }
     }
@@ -246,8 +251,8 @@ public class MultiStatementCypherSubGraphExporter {
         while (labels.hasNext()) {
             Label next = labels.next();
             String labelName = next.name();
-            if (uniqueConstraints.containsKey(labelName) && node.hasProperty(uniqueConstraints.get(labelName)))
-                uniqueFound = true;
+            uniqueFound = CypherFormatterUtils.isUniqueLabelFound(node, uniqueConstraints, labelName);
+
         }
         if (!uniqueFound) {
             artificialUniques++;

--- a/src/main/java/apoc/export/cypher/MultiStatementCypherSubGraphExporter.java
+++ b/src/main/java/apoc/export/cypher/MultiStatementCypherSubGraphExporter.java
@@ -163,8 +163,7 @@ public class MultiStatementCypherSubGraphExporter {
             String label = index.getLabel().name();
             Iterable<String> props = index.getPropertyKeys();
             if (index.isConstraintIndex()) {
-                String cypher;
-                cypher = this.cypherFormat.statementForConstraint(label, props);
+                String cypher = this.cypherFormat.statementForConstraint(label, props);
                 if (cypher != null && !"".equals(cypher)) {
                     result.add(cypher);
                 }

--- a/src/main/java/apoc/export/cypher/formatter/AbstractCypherFormatter.java
+++ b/src/main/java/apoc/export/cypher/formatter/AbstractCypherFormatter.java
@@ -2,9 +2,12 @@ package apoc.export.cypher.formatter;
 
 import org.neo4j.graphdb.Node;
 import org.neo4j.graphdb.Relationship;
+import org.neo4j.helpers.collection.Iterables;
 
 import java.util.Map;
 import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
 
 import static apoc.export.cypher.formatter.CypherFormatterUtils.Q_UNIQUE_ID_LABEL;
 import static apoc.export.cypher.formatter.CypherFormatterUtils.UNIQUE_ID_PROP;
@@ -16,6 +19,8 @@ import static apoc.export.cypher.formatter.CypherFormatterUtils.quote;
  * @since 16-06-2017
  */
 abstract class AbstractCypherFormatter implements CypherFormatter {
+
+	private static final String STATEMENT_CONSTRAINTS = "CREATE CONSTRAINT ON (node:%s) ASSERT (%s) %s;";
 
 	@Override
 	public String statementForCleanUp(int batchSize) {
@@ -30,11 +35,16 @@ abstract class AbstractCypherFormatter implements CypherFormatter {
 	}
 
 	@Override
-	public String statementForConstraint(String label, String key) {
-		return "CREATE CONSTRAINT ON (node:" + CypherFormatterUtils.quote(label) + ") ASSERT node." + CypherFormatterUtils.quote(key) + " IS UNIQUE;";
+	public String statementForConstraint(String label, Iterable<String> keys) {
+
+		String keysString = StreamSupport.stream(keys.spliterator(), false)
+				.map(key -> "node." + CypherFormatterUtils.quote(key))
+				.collect(Collectors.joining(", "));
+
+		return  String.format(STATEMENT_CONSTRAINTS, CypherFormatterUtils.quote(label), keysString, Iterables.count(keys) > 1 ? "IS NODE KEY" : "IS UNIQUE");
 	}
 
-	protected String mergeStatementForNode(CypherFormat cypherFormat, Node node, Map<String, String> uniqueConstraints, Set<String> indexedProperties, Set<String> indexNames) {
+	protected String mergeStatementForNode(CypherFormat cypherFormat, Node node, Map<String, Set<String>> uniqueConstraints, Set<String> indexedProperties, Set<String> indexNames) {
 		StringBuilder result = new StringBuilder(1000);
 		result.append("MERGE ");
 		result.append(CypherFormatterUtils.formatNodeLookup("n", node, uniqueConstraints, indexNames));
@@ -52,7 +62,7 @@ abstract class AbstractCypherFormatter implements CypherFormatter {
 		return result.toString();
 	}
 
-	public String mergeStatementForRelationship(CypherFormat cypherFormat, Relationship relationship, Map<String, String> uniqueConstraints, Set<String> indexedProperties) {
+	public String mergeStatementForRelationship(CypherFormat cypherFormat, Relationship relationship, Map<String, Set<String>> uniqueConstraints, Set<String> indexedProperties) {
 		StringBuilder result = new StringBuilder(1000);
 		result.append("MATCH ");
 		result.append(CypherFormatterUtils.formatNodeLookup("n1", relationship.getStartNode(), uniqueConstraints, indexedProperties));

--- a/src/main/java/apoc/export/cypher/formatter/AddStructureCypherFormatter.java
+++ b/src/main/java/apoc/export/cypher/formatter/AddStructureCypherFormatter.java
@@ -14,12 +14,12 @@ import java.util.Set;
 public class AddStructureCypherFormatter extends AbstractCypherFormatter implements CypherFormatter {
 
 	@Override
-	public String statementForNode(Node node, Map<String, String> uniqueConstraints, Set<String> indexedProperties, Set<String> indexNames) {
+	public String statementForNode(Node node, Map<String, Set<String>> uniqueConstraints, Set<String> indexedProperties, Set<String> indexNames) {
 		return super.mergeStatementForNode(CypherFormat.ADD_STRUCTURE, node, uniqueConstraints, indexedProperties, indexNames);
 	}
 
 	@Override
-	public String statementForRelationship(Relationship relationship,  Map<String, String> uniqueConstraints, Set<String> indexedProperties) {
+	public String statementForRelationship(Relationship relationship, Map<String, Set<String>> uniqueConstraints, Set<String> indexedProperties) {
 		return new CreateCypherFormatter().statementForRelationship(relationship, uniqueConstraints, indexedProperties);
 	}
 
@@ -34,7 +34,7 @@ public class AddStructureCypherFormatter extends AbstractCypherFormatter impleme
 	}
 
 	@Override
-	public String statementForConstraint(String label, String key) {
+	public String statementForConstraint(String label, Iterable<String> keys) {
 		return "";
 	}
 }

--- a/src/main/java/apoc/export/cypher/formatter/CreateCypherFormatter.java
+++ b/src/main/java/apoc/export/cypher/formatter/CreateCypherFormatter.java
@@ -14,7 +14,7 @@ import java.util.Set;
 public class CreateCypherFormatter extends AbstractCypherFormatter implements CypherFormatter {
 
 	@Override
-    public String statementForNode(Node node, Map<String, String> uniqueConstraints, Set<String> indexedProperties, Set<String> indexNames) {
+    public String statementForNode(Node node, Map<String, Set<String>> uniqueConstraints, Set<String> indexedProperties, Set<String> indexNames) {
 		StringBuilder result = new StringBuilder(100);
 		result.append("CREATE (");
 		String labels = CypherFormatterUtils.formatAllLabels(node, uniqueConstraints, indexNames);
@@ -31,7 +31,7 @@ public class CreateCypherFormatter extends AbstractCypherFormatter implements Cy
 	}
 
 	@Override
-    public String statementForRelationship(Relationship relationship,  Map<String, String> uniqueConstraints, Set<String> indexedProperties) {
+    public String statementForRelationship(Relationship relationship, Map<String, Set<String>> uniqueConstraints, Set<String> indexedProperties) {
 		StringBuilder result = new StringBuilder(100);
 		result.append("MATCH ");
 		result.append(CypherFormatterUtils.formatNodeLookup("n1", relationship.getStartNode(), uniqueConstraints, indexedProperties));

--- a/src/main/java/apoc/export/cypher/formatter/CypherFormatter.java
+++ b/src/main/java/apoc/export/cypher/formatter/CypherFormatter.java
@@ -13,13 +13,13 @@ import java.util.Set;
  */
 public interface CypherFormatter {
 
-	String statementForNode(Node node, Map<String, String> uniqueConstraints, Set<String> indexedProperties, Set<String> indexNames);
+	String statementForNode(Node node, Map<String, Set<String>> uniqueConstraints, Set<String> indexedProperties, Set<String> indexNames);
 
-	String statementForRelationship(Relationship relationship,  Map<String, String> uniqueConstraints, Set<String> indexedProperties);
+	String statementForRelationship(Relationship relationship, Map<String, Set<String>> uniqueConstraints, Set<String> indexedProperties);
 
 	String statementForIndex(String label, Iterable<String> keys);
 
-	String statementForConstraint(String label, String key);
+	String statementForConstraint(String label, Iterable<String> keys);
 
 	String statementForCleanUp(int batchSize);
 }

--- a/src/main/java/apoc/export/cypher/formatter/CypherFormatterUtils.java
+++ b/src/main/java/apoc/export/cypher/formatter/CypherFormatterUtils.java
@@ -29,7 +29,7 @@ public class CypherFormatterUtils {
 
     // ---- node id ----
 
-    public static  String formatNodeLookup(String id, Node node, Map<String, String> uniqueConstraints, Set<String> indexNames) {
+    public static  String formatNodeLookup(String id, Node node, Map<String, Set<String>> uniqueConstraints, Set<String> indexNames) {
         StringBuilder result = new StringBuilder(100);
         result.append("(");
         result.append(id);
@@ -51,17 +51,19 @@ public class CypherFormatterUtils {
         return result.toString();
     }
 
-    private static Map<String, Object> getNodeIdProperties(Node node, Map<String, String> uniqueConstraints) {
+    private static Map<String, Object> getNodeIdProperties(Node node, Map<String, Set<String>> uniqueConstraints) {
         Map<String, Object> nodeIdProperties = new LinkedHashMap<>();
         boolean uniqueLabelFound = false;
         List<String> list = getLabelsSorted(node);
 
         for (String labelName : list) {
-            String prop = uniqueConstraints.get(labelName);
-            if (prop != null && node.hasProperty(prop)) {
-                uniqueLabelFound = true;
-                nodeIdProperties.put(prop, node.getProperty(prop));
+            uniqueLabelFound = isUniqueLabelFound(node, uniqueConstraints, labelName);
+            if (!uniqueLabelFound) {
+                continue;
             }
+            uniqueConstraints.get(labelName).forEach(prop -> {
+                nodeIdProperties.put(prop, node.getProperty(prop));
+            });
         }
         if (!uniqueLabelFound) {
             nodeIdProperties.put(UNIQUE_ID_PROP, node.getId());
@@ -71,15 +73,13 @@ public class CypherFormatterUtils {
 
     // ---- labels ----
 
-    public static String formatAllLabels(Node node, Map<String, String> uniqueConstraints, Set<String> indexNames) {
+    public static String formatAllLabels(Node node, Map<String, Set<String>> uniqueConstraints, Set<String> indexNames) {
         StringBuilder result = new StringBuilder(100);
         boolean uniqueLabelFound = false;
         List<String> list = getLabelsSorted(node);
 
         for (String labelName : list) {
-            String prop = uniqueConstraints.get(labelName);
-            if (prop != null && node.hasProperty(prop))
-                uniqueLabelFound = true;
+            uniqueLabelFound = isUniqueLabelFound(node, uniqueConstraints, labelName);
             if (indexNames != null && indexNames.contains(labelName))
                 result.insert(0, label(labelName));
             else
@@ -91,13 +91,12 @@ public class CypherFormatterUtils {
         return result.toString();
     }
 
-    public static String formatNotUniqueLabels(String id, Node node, Map<String, String> uniqueConstraints) {
+    public static String formatNotUniqueLabels(String id, Node node, Map<String, Set<String>> uniqueConstraints) {
         StringBuilder result = new StringBuilder(100);
         List<String> list = getLabelsSorted(node);
 
         for (String labelName : list) {
-            String prop = uniqueConstraints.get(labelName);
-            if (!node.hasProperty(prop)) {
+            if (!isUniqueLabelFound(node, uniqueConstraints, labelName)) {
                 result.append(", ");
                 result.append(id);
                 result.append(label(labelName));
@@ -106,21 +105,23 @@ public class CypherFormatterUtils {
         return result.length() > 0 ? result.substring(2) : "";
     }
 
-    private static String getNodeIdLabels(Node node, Map<String, String> uniqueConstraints, Set<String> indexNames) {
+    private static String getNodeIdLabels(Node node, Map<String, Set<String>> uniqueConstraints, Set<String> indexNames) {
         StringBuilder result = new StringBuilder(100);
         boolean uniqueLabelFound = false;
         List<String> list = getLabelsSorted(node);
 
         for (String labelName : list) {
-            String prop = uniqueConstraints.get(labelName);
-            if (prop != null && node.hasProperty(prop)) {
-                uniqueLabelFound = true;
-                if (indexNames != null && indexNames.contains(labelName)) {
-                    result.insert(0, label(labelName));
-                }
-                else
-                    result.append(label(labelName));
+            uniqueLabelFound = isUniqueLabelFound(node, uniqueConstraints, labelName);
+
+            if (!uniqueLabelFound) {
+                continue;
             }
+
+            if (indexNames != null && indexNames.contains(labelName)) {
+                result.insert(0, label(labelName));
+            }
+            else
+                result.append(label(labelName));
         }
         if (!uniqueLabelFound) {
             result.append(label(UNIQUE_ID_LABEL));
@@ -128,9 +129,18 @@ public class CypherFormatterUtils {
         return result.toString();
     }
 
+    public static boolean isUniqueLabelFound(Node node, Map<String, Set<String>> uniqueConstraints, String labelName) {
+        if (uniqueConstraints.containsKey(labelName)) {
+            Set<String> nodeUniqueConstraint = uniqueConstraints.get(labelName);
+            return nodeUniqueConstraint.stream().allMatch(node::hasProperty);
+        } else {
+            return false;
+        }
+    }
+
     // ---- properties ----
 
-    public static String formatNodeProperties(String id, Node node, Map<String, String> uniqueConstraints, Set<String> indexNames, boolean jsonStyle) {
+    public static String formatNodeProperties(String id, Node node, Map<String, Set<String>> uniqueConstraints, Set<String> indexNames, boolean jsonStyle) {
         StringBuilder result = formatProperties(id, node.getAllProperties(), jsonStyle);
         if (getNodeIdLabels(node, uniqueConstraints, indexNames).endsWith(label(UNIQUE_ID_LABEL))) {
             result.append(", ");
@@ -144,7 +154,7 @@ public class CypherFormatterUtils {
         return result.length() > 0 ? result.substring(2) : "";
     }
 
-    public static String formatNotUniqueProperties(String id, Node node, Map<String, String> uniqueConstraints, Set<String> indexedProperties, boolean jsonStyle) {
+    public static String formatNotUniqueProperties(String id, Node node, Map<String, Set<String>> uniqueConstraints, Set<String> indexedProperties, boolean jsonStyle) {
         Map<String, Object> properties = new LinkedHashMap<>();
         List<String> keys = Iterables.asList(node.getPropertyKeys());
         Collections.sort(keys);

--- a/src/main/java/apoc/export/cypher/formatter/UpdateAllCypherFormatter.java
+++ b/src/main/java/apoc/export/cypher/formatter/UpdateAllCypherFormatter.java
@@ -14,12 +14,12 @@ import java.util.Set;
 public class UpdateAllCypherFormatter extends AbstractCypherFormatter implements CypherFormatter {
 
 	@Override
-	public String statementForNode(Node node, Map<String, String> uniqueConstraints, Set<String> indexedProperties, Set<String> indexNames) {
+	public String statementForNode(Node node, Map<String, Set<String>> uniqueConstraints, Set<String> indexedProperties, Set<String> indexNames) {
 	    return super.mergeStatementForNode(CypherFormat.UPDATE_ALL, node, uniqueConstraints, indexedProperties, indexNames);
 	}
 
 	@Override
-	public String statementForRelationship(Relationship relationship,  Map<String, String> uniqueConstraints, Set<String> indexedProperties) {
+	public String statementForRelationship(Relationship relationship, Map<String, Set<String>> uniqueConstraints, Set<String> indexedProperties) {
         return super.mergeStatementForRelationship(CypherFormat.UPDATE_ALL, relationship, uniqueConstraints, indexedProperties);
 	}
 }

--- a/src/main/java/apoc/export/cypher/formatter/UpdateStructureCypherFormatter.java
+++ b/src/main/java/apoc/export/cypher/formatter/UpdateStructureCypherFormatter.java
@@ -14,12 +14,12 @@ import java.util.Set;
 public class UpdateStructureCypherFormatter extends AbstractCypherFormatter implements CypherFormatter {
 
 	@Override
-	public String statementForNode(Node node, Map<String, String> uniqueConstraints, Set<String> indexedProperties, Set<String> indexNames) {
+	public String statementForNode(Node node, Map<String, Set<String>> uniqueConstraints, Set<String> indexedProperties, Set<String> indexNames) {
 		return "";
 	}
 
 	@Override
-	public String statementForRelationship(Relationship relationship,  Map<String, String> uniqueConstraints, Set<String> indexedProperties) {
+	public String statementForRelationship(Relationship relationship, Map<String, Set<String>> uniqueConstraints, Set<String> indexedProperties) {
         return super.mergeStatementForRelationship(CypherFormat.UPDATE_STRUCTURE, relationship, uniqueConstraints, indexedProperties);
 	}
 
@@ -34,7 +34,7 @@ public class UpdateStructureCypherFormatter extends AbstractCypherFormatter impl
 	}
 
 	@Override
-	public String statementForConstraint(String label, String key) {
+	public String statementForConstraint(String label, Iterable<String> key) {
 		return "";
 	}
 }

--- a/src/test/java/apoc/export/cypher/ExportCypherTest.java
+++ b/src/test/java/apoc/export/cypher/ExportCypherTest.java
@@ -5,7 +5,9 @@ import apoc.util.TestUtil;
 import apoc.util.Util;
 import org.junit.After;
 import org.junit.Before;
+import org.junit.Rule;
 import org.junit.Test;
+import org.junit.rules.TestName;
 import org.neo4j.graphdb.GraphDatabaseService;
 import org.neo4j.graphdb.factory.GraphDatabaseSettings;
 import org.neo4j.test.TestGraphDatabaseFactory;
@@ -19,6 +21,7 @@ import java.util.Scanner;
 import static apoc.export.cypher.ExportCypherTest.ExportCypherResults.*;
 import static apoc.export.util.ExportFormat.*;
 import static apoc.util.MapUtil.map;
+import static apoc.util.TestUtil.apocEnterpriseGraphDatabaseBuilder;
 import static org.junit.Assert.*;
 
 /**
@@ -35,22 +38,41 @@ public class ExportCypherTest {
         directory.mkdirs();
     }
 
+    @Rule
+    public TestName testName = new TestName();
+
+    private static final String TEST_WITH_ENTERPRISE_DB = "WithEnterpriseDB";
+
     @Before
     public void setUp() throws Exception {
-        db = new TestGraphDatabaseFactory().newImpermanentDatabaseBuilder().setConfig(GraphDatabaseSettings.load_csv_file_url_root, directory.getAbsolutePath())
-                .setConfig("apoc.export.file.enabled", "true").newGraphDatabase();
-        TestUtil.registerProcedure(db, ExportCypher.class, Graphs.class);
-        db.execute("CREATE INDEX ON :Foo(name)").close();
-        db.execute("CREATE INDEX ON :Bar(first_name, last_name)").close();
-        db.execute("CREATE CONSTRAINT ON (b:Bar) ASSERT b.name IS UNIQUE").close();
-        db.execute("CREATE (f:Foo {name:'foo', born:date('2018-10-31')})-[:KNOWS {since:2016}]->(b:Bar {name:'bar',age:42}),(c:Bar {age:12})").close();
+
+        if(testName.getMethodName().endsWith(TEST_WITH_ENTERPRISE_DB)) {
+            db =  apocEnterpriseGraphDatabaseBuilder().setConfig(GraphDatabaseSettings.load_csv_file_url_root, directory.getAbsolutePath())
+                    .setConfig("apoc.export.file.enabled", "true").newGraphDatabase();
+            TestUtil.registerProcedure(db, ExportCypher.class, Graphs.class);
+            db.execute("CREATE CONSTRAINT ON (t:Person) ASSERT (t.name, t.surname) IS NODE KEY;").close();
+            db.execute("CREATE (a:Person {name: 'John', surname: 'Snow'}) " +
+                    "CREATE (b:Person {name: 'Matt', surname: 'Jackson'}) " +
+                    "CREATE (c:Person {name: 'Jenny', surname: 'White'}) " +
+                    "CREATE (d:Person {name: 'Susan', surname: 'Brown'}) " +
+                    "CREATE (e:Person {name: 'Tom', surname: 'Taylor'})" +
+                    "CREATE (a)-[:KNOWS]->(b);").close();
+        } else {
+            db = new TestGraphDatabaseFactory().newImpermanentDatabaseBuilder().setConfig(GraphDatabaseSettings.load_csv_file_url_root, directory.getAbsolutePath())
+                    .setConfig("apoc.export.file.enabled", "true").newGraphDatabase();
+            TestUtil.registerProcedure(db, ExportCypher.class, Graphs.class);
+            db.execute("CREATE INDEX ON :Foo(name)").close();
+            db.execute("CREATE INDEX ON :Bar(first_name, last_name)").close();
+            db.execute("CREATE CONSTRAINT ON (b:Bar) ASSERT b.name IS UNIQUE").close();
+            db.execute("CREATE (f:Foo {name:'foo', born:date('2018-10-31')})-[:KNOWS {since:2016}]->(b:Bar {name:'bar',age:42}),(c:Bar {age:12})").close();
+        }
+
     }
 
     @After
     public void tearDown() {
         db.shutdown();
     }
-
 
     @Test
     public void testExportAllCypherResults() {
@@ -341,6 +363,33 @@ public class ExportCypherTest {
         assertEquals(EXPECTED_CYPHER_LABELS_ASCENDEND, readFile(output));
     }
 
+    @Test
+    public void testExportWithCompoundConstraintCypherShellWithEnterpriseDB() throws Exception {
+        File output = new File(directory, "testCypherShellWithCompoundConstraint.cypher");
+        TestUtil.testCall(db, "CALL apoc.export.cypher.all({file},{config});\n",
+                map("file", output.getAbsolutePath(), "config", Util.map("format", "cypher-shell")), (r) -> {
+                });
+        assertEquals(EXPECTED_CYPHER_SHELL_WITH_COMPOUND_CONSTRAINT, readFile(output));
+    }
+
+    @Test
+    public void testExportWithCompoundConstraintPlainWithEnterpriseDB() throws Exception {
+        File output = new File(directory, "testPlainFormatWithCompoundConstraint.cypher");
+        TestUtil.testCall(db, "CALL apoc.export.cypher.all({file},{config});\n",
+                map("file", output.getAbsolutePath(), "config", Util.map("format", "plain")), (r) -> {
+                });
+        assertEquals(EXPECTED_PLAIN_FORMAT_WITH_COMPOUND_CONSTRAINT, readFile(output));
+    }
+
+    @Test
+    public void testExportWithCompoundConstraintNeo4jShellWithEnterpriseDB() throws Exception {
+        File output = new File(directory, "testNeo4jShellWithCompoundConstraint.cypher");
+        TestUtil.testCall(db, "CALL apoc.export.cypher.all({file},{config});\n",
+                map("file", output.getAbsolutePath(), "config", Util.map("format", "neo4j-shell")), (r) -> {
+                });
+        assertEquals(EXPECTED_NEO4J_SHELL_WITH_COMPOUND_CONSTRAINT, readFile(output));
+    }
+
     static class ExportCypherResults {
 
         static final String EXPECTED_NODES = String.format("BEGIN%n" +
@@ -364,8 +413,8 @@ public class ExportCypherTest {
         static final String EXPECTED_SCHEMA = String.format("BEGIN%n" +
                 "CREATE INDEX ON :`Bar`(`first_name`,`last_name`);%n" +
                 "CREATE INDEX ON :`Foo`(`name`);%n" +
-                "CREATE CONSTRAINT ON (node:`Bar`) ASSERT node.`name` IS UNIQUE;%n" +
-                "CREATE CONSTRAINT ON (node:`UNIQUE IMPORT LABEL`) ASSERT node.`UNIQUE IMPORT ID` IS UNIQUE;%n" +
+                "CREATE CONSTRAINT ON (node:`Bar`) ASSERT (node.`name`) IS UNIQUE;%n" +
+                "CREATE CONSTRAINT ON (node:`UNIQUE IMPORT LABEL`) ASSERT (node.`UNIQUE IMPORT ID`) IS UNIQUE;%n" +
                 "COMMIT%n" +
                 "SCHEMA AWAIT%n");
 
@@ -392,7 +441,7 @@ public class ExportCypherTest {
                 "MATCH (n:`UNIQUE IMPORT LABEL`)  WITH n LIMIT 20000 REMOVE n:`UNIQUE IMPORT LABEL` REMOVE n.`UNIQUE IMPORT ID`;%n" +
                 "COMMIT%n" +
                 "BEGIN%n" +
-                "DROP CONSTRAINT ON (node:`UNIQUE IMPORT LABEL`) ASSERT node.`UNIQUE IMPORT ID` IS UNIQUE;%n" +
+                "DROP CONSTRAINT ON (node:`UNIQUE IMPORT LABEL`) ASSERT (node.`UNIQUE IMPORT ID`) IS UNIQUE;%n" +
                 "COMMIT%n");
 
         static final String EXPECTED_CLEAN_UP_EMPTY = String.format("BEGIN%n" +
@@ -403,7 +452,7 @@ public class ExportCypherTest {
         static final String EXPECTED_ONLY_SCHEMA_NEO4J_SHELL = String.format("BEGIN%n" +
                 "CREATE INDEX ON :`Bar`(`first_name`,`last_name`);%n" +
                 "CREATE INDEX ON :`Foo`(`name`);%n" +
-                "CREATE CONSTRAINT ON (node:`Bar`) ASSERT node.`name` IS UNIQUE;%n" +
+                "CREATE CONSTRAINT ON (node:`Bar`) ASSERT (node.`name`) IS UNIQUE;%n" +
                 "COMMIT%n" +
                 "SCHEMA AWAIT%n");
 
@@ -413,8 +462,8 @@ public class ExportCypherTest {
                 "COMMIT%n" +
                 "BEGIN%n" +
                 "CREATE INDEX ON :`Bar`(`first_name`,`last_name`);%n" +
-                "CREATE CONSTRAINT ON (node:`Bar`) ASSERT node.`name` IS UNIQUE;%n" +
-                "CREATE CONSTRAINT ON (node:`UNIQUE IMPORT LABEL`) ASSERT node.`UNIQUE IMPORT ID` IS UNIQUE;%n" +
+                "CREATE CONSTRAINT ON (node:`Bar`) ASSERT (node.`name`) IS UNIQUE;%n" +
+                "CREATE CONSTRAINT ON (node:`UNIQUE IMPORT LABEL`) ASSERT (node.`UNIQUE IMPORT ID`) IS UNIQUE;%n" +
                 "COMMIT%n" +
                 "SCHEMA AWAIT%n" +
                 "BEGIN%n" +
@@ -424,7 +473,7 @@ public class ExportCypherTest {
                 "MATCH (n:`UNIQUE IMPORT LABEL`)  WITH n LIMIT 20000 REMOVE n:`UNIQUE IMPORT LABEL` REMOVE n.`UNIQUE IMPORT ID`;%n" +
                 "COMMIT%n" +
                 "BEGIN%n" +
-                "DROP CONSTRAINT ON (node:`UNIQUE IMPORT LABEL`) ASSERT node.`UNIQUE IMPORT ID` IS UNIQUE;%n" +
+                "DROP CONSTRAINT ON (node:`UNIQUE IMPORT LABEL`) ASSERT (node.`UNIQUE IMPORT ID`) IS UNIQUE;%n" +
                 "COMMIT%n");
 
         static final String EXPECTED_CYPHER_DATE = String.format("BEGIN%n" +
@@ -433,8 +482,8 @@ public class ExportCypherTest {
                 "COMMIT%n" +
                 "BEGIN%n" +
                 "CREATE INDEX ON :`Bar`(`first_name`,`last_name`);%n" +
-                "CREATE CONSTRAINT ON (node:`Bar`) ASSERT node.`name` IS UNIQUE;%n" +
-                "CREATE CONSTRAINT ON (node:`UNIQUE IMPORT LABEL`) ASSERT node.`UNIQUE IMPORT ID` IS UNIQUE;%n" +
+                "CREATE CONSTRAINT ON (node:`Bar`) ASSERT (node.`name`) IS UNIQUE;%n" +
+                "CREATE CONSTRAINT ON (node:`UNIQUE IMPORT LABEL`) ASSERT (node.`UNIQUE IMPORT ID`) IS UNIQUE;%n" +
                 "COMMIT%n" +
                 "SCHEMA AWAIT%n" +
                 "BEGIN%n" +
@@ -444,7 +493,7 @@ public class ExportCypherTest {
                 "MATCH (n:`UNIQUE IMPORT LABEL`)  WITH n LIMIT 20000 REMOVE n:`UNIQUE IMPORT LABEL` REMOVE n.`UNIQUE IMPORT ID`;%n" +
                 "COMMIT%n" +
                 "BEGIN%n" +
-                "DROP CONSTRAINT ON (node:`UNIQUE IMPORT LABEL`) ASSERT node.`UNIQUE IMPORT ID` IS UNIQUE;%n" +
+                "DROP CONSTRAINT ON (node:`UNIQUE IMPORT LABEL`) ASSERT (node.`UNIQUE IMPORT ID`) IS UNIQUE;%n" +
                 "COMMIT%n");
 
         static final String EXPECTED_CYPHER_TIME = String.format("BEGIN%n" +
@@ -453,8 +502,8 @@ public class ExportCypherTest {
                 "COMMIT%n" +
                 "BEGIN%n" +
                 "CREATE INDEX ON :`Bar`(`first_name`,`last_name`);%n" +
-                "CREATE CONSTRAINT ON (node:`Bar`) ASSERT node.`name` IS UNIQUE;%n" +
-                "CREATE CONSTRAINT ON (node:`UNIQUE IMPORT LABEL`) ASSERT node.`UNIQUE IMPORT ID` IS UNIQUE;%n" +
+                "CREATE CONSTRAINT ON (node:`Bar`) ASSERT (node.`name`) IS UNIQUE;%n" +
+                "CREATE CONSTRAINT ON (node:`UNIQUE IMPORT LABEL`) ASSERT (node.`UNIQUE IMPORT ID`) IS UNIQUE;%n" +
                 "COMMIT%n" +
                 "SCHEMA AWAIT%n" +
                 "BEGIN%n" +
@@ -464,7 +513,7 @@ public class ExportCypherTest {
                 "MATCH (n:`UNIQUE IMPORT LABEL`)  WITH n LIMIT 20000 REMOVE n:`UNIQUE IMPORT LABEL` REMOVE n.`UNIQUE IMPORT ID`;%n" +
                 "COMMIT%n" +
                 "BEGIN%n" +
-                "DROP CONSTRAINT ON (node:`UNIQUE IMPORT LABEL`) ASSERT node.`UNIQUE IMPORT ID` IS UNIQUE;%n" +
+                "DROP CONSTRAINT ON (node:`UNIQUE IMPORT LABEL`) ASSERT (node.`UNIQUE IMPORT ID`) IS UNIQUE;%n" +
                 "COMMIT%n");
 
         static final String EXPECTED_CYPHER_DURATION = String.format("BEGIN%n" +
@@ -473,8 +522,8 @@ public class ExportCypherTest {
                 "COMMIT%n" +
                 "BEGIN%n" +
                 "CREATE INDEX ON :`Bar`(`first_name`,`last_name`);%n" +
-                "CREATE CONSTRAINT ON (node:`Bar`) ASSERT node.`name` IS UNIQUE;%n" +
-                "CREATE CONSTRAINT ON (node:`UNIQUE IMPORT LABEL`) ASSERT node.`UNIQUE IMPORT ID` IS UNIQUE;%n" +
+                "CREATE CONSTRAINT ON (node:`Bar`) ASSERT (node.`name`) IS UNIQUE;%n" +
+                "CREATE CONSTRAINT ON (node:`UNIQUE IMPORT LABEL`) ASSERT (node.`UNIQUE IMPORT ID`) IS UNIQUE;%n" +
                 "COMMIT%n" +
                 "SCHEMA AWAIT%n" +
                 "BEGIN%n" +
@@ -484,21 +533,21 @@ public class ExportCypherTest {
                 "MATCH (n:`UNIQUE IMPORT LABEL`)  WITH n LIMIT 20000 REMOVE n:`UNIQUE IMPORT LABEL` REMOVE n.`UNIQUE IMPORT ID`;%n" +
                 "COMMIT%n" +
                 "BEGIN%n" +
-                "DROP CONSTRAINT ON (node:`UNIQUE IMPORT LABEL`) ASSERT node.`UNIQUE IMPORT ID` IS UNIQUE;%n" +
+                "DROP CONSTRAINT ON (node:`UNIQUE IMPORT LABEL`) ASSERT (node.`UNIQUE IMPORT ID`) IS UNIQUE;%n" +
                 "COMMIT%n");
 
         static final String EXPECTED_CYPHER_LABELS_ASCENDEND = String.format("BEGIN%n" +
                 "CREATE (:`User`:`User0`:`User1`:`User12`:`UNIQUE IMPORT LABEL` {`name`:\"Alan\", `UNIQUE IMPORT ID`:20});%n" +
                 "COMMIT%n" +
                 "BEGIN%n" +
-                "CREATE CONSTRAINT ON (node:`UNIQUE IMPORT LABEL`) ASSERT node.`UNIQUE IMPORT ID` IS UNIQUE;%n" +
+                "CREATE CONSTRAINT ON (node:`UNIQUE IMPORT LABEL`) ASSERT (node.`UNIQUE IMPORT ID`) IS UNIQUE;%n" +
                 "COMMIT%n" +
                 "SCHEMA AWAIT%n" +
                 "BEGIN%n" +
                 "MATCH (n:`UNIQUE IMPORT LABEL`)  WITH n LIMIT 20000 REMOVE n:`UNIQUE IMPORT LABEL` REMOVE n.`UNIQUE IMPORT ID`;%n" +
                 "COMMIT%n" +
                 "BEGIN%n" +
-                "DROP CONSTRAINT ON (node:`UNIQUE IMPORT LABEL`) ASSERT node.`UNIQUE IMPORT ID` IS UNIQUE;%n" +
+                "DROP CONSTRAINT ON (node:`UNIQUE IMPORT LABEL`) ASSERT (node.`UNIQUE IMPORT ID`) IS UNIQUE;%n" +
                 "COMMIT%n");
 
         static final String EXPECTED_NEO4J_SHELL = EXPECTED_NODES + EXPECTED_SCHEMA + EXPECTED_RELATIONSHIPS + EXPECTED_CLEAN_UP;
@@ -518,6 +567,37 @@ public class ExportCypherTest {
         static final String EXPECTED_ONLY_SCHEMA_CYPHER_SHELL = EXPECTED_ONLY_SCHEMA_NEO4J_SHELL.replace(NEO4J_SHELL.begin(), CYPHER_SHELL.begin())
                 .replace(NEO4J_SHELL.commit(), CYPHER_SHELL.commit()).replace(NEO4J_SHELL.schemaAwait(), CYPHER_SHELL.schemaAwait()) + EXPECTED_INDEXES_AWAIT;
 
+        static final String EXPECTED_NODES_COMPOUND_CONSTRAINT = String.format("BEGIN%n" +
+                "CREATE (:`Person` {`name`:\"John\", `surname`:\"Snow\"});%n" +
+                "CREATE (:`Person` {`name`:\"Matt\", `surname`:\"Jackson\"});%n" +
+                "CREATE (:`Person` {`name`:\"Jenny\", `surname`:\"White\"});%n" +
+                "CREATE (:`Person` {`name`:\"Susan\", `surname`:\"Brown\"});%n" +
+                "CREATE (:`Person` {`name`:\"Tom\", `surname`:\"Taylor\"});%n" +
+                "COMMIT%n");
+
+        static final String EXPECTED_SCHEMA_COMPOUND_CONSTRAINT = String.format("BEGIN%n" +
+                "CREATE CONSTRAINT ON (node:`Person`) ASSERT (node.`name`, node.`surname`) IS NODE KEY;%n" +
+                "COMMIT%n" +
+                "SCHEMA AWAIT%n");
+
+        static final String EXPECTED_RELATIONSHIP_COMPOUND_CONSTRAINT = String.format(("BEGIN%n" +
+                "MATCH (n1:`Person`{`surname`:\"Snow\", `name`:\"John\"}), (n2:`Person`{`surname`:\"Jackson\", `name`:\"Matt\"}) CREATE (n1)-[r:`KNOWS`]->(n2);%n" +
+                "COMMIT%n"));
+
+        static final String EXPECTED_INDEX_AWAIT_COMPOUND_CONSTRAINT =  String.format("CALL db.awaitIndex(':`Person`(`name`,`surname`)');%n");
+
+        static final String EXPECTED_NEO4J_SHELL_WITH_COMPOUND_CONSTRAINT = EXPECTED_NODES_COMPOUND_CONSTRAINT + EXPECTED_SCHEMA_COMPOUND_CONSTRAINT + EXPECTED_RELATIONSHIP_COMPOUND_CONSTRAINT;
+
+        static final String EXPECTED_CYPHER_SHELL_WITH_COMPOUND_CONSTRAINT = EXPECTED_NEO4J_SHELL_WITH_COMPOUND_CONSTRAINT
+                .replace(NEO4J_SHELL.begin(), CYPHER_SHELL.begin())
+                .replace(NEO4J_SHELL.commit(), CYPHER_SHELL.commit())
+                .replace(NEO4J_SHELL.schemaAwait(), EXPECTED_INDEX_AWAIT_COMPOUND_CONSTRAINT)
+                .replace(NEO4J_SHELL.schemaAwait(), CYPHER_SHELL.schemaAwait());
+
+        static final String EXPECTED_PLAIN_FORMAT_WITH_COMPOUND_CONSTRAINT = EXPECTED_NEO4J_SHELL_WITH_COMPOUND_CONSTRAINT
+                .replace(NEO4J_SHELL.begin(), PLAIN_FORMAT.begin())
+                .replace(NEO4J_SHELL.commit(), PLAIN_FORMAT.commit())
+                .replace(NEO4J_SHELL.schemaAwait(), PLAIN_FORMAT.schemaAwait());
     }
 
 }

